src/scipp/core/binning.py

```diff
@@ -0,0 +1,528 @@
+                        num=arg + 1).to(dtype=start.dtype, copy=False)
+    start = coord.min()
+    step = arg.to(dtype=start.dtype, unit=start.unit)
+    stop = _upper_bound(coord) + step
+    # See #2639, need explicit dtype currently
+    return arange(name, start, stop, step=step, dtype=start.dtype)
+
+
+def _make_edges(x: Union[_cpp.DataArray,
+                         _cpp.Dataset], arg_dict: Dict[str, Union[int, Variable]],
+                kwargs: Dict[str, Union[int, Variable]]) -> List[Variable]:
+    if arg_dict is not None:
+        kwargs = dict(**arg_dict, **kwargs)
+    return {name: _parse_coords_arg(x, name, arg) for name, arg in kwargs.items()}
+
+
+def _find_replaced_dims(x, dims):
+    if x.bins is None:
+        return []
+    erase = set()
+    for dim in dims:
+        if (coord := x.meta.get(dim)) is not None:
+            if dim not in coord.dims:
+                erase = erase.union(coord.dims)
+    return [dim for dim in erase if dim not in dims]
+
+
+def hist(x: Union[_cpp.DataArray, _cpp.Dataset],
+         arg_dict: Optional[Dict[str, Union[int, Variable]]] = None,
+         /,
+         **kwargs: Union[int, Variable]) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    """Compute a histogram.
+
+    Bin edges can be specified in three ways: (1) When an integer is provided,
+    a 'linspace' with this requested number of bins is created, based on the min
+    and max of the corresponding coordinate. (2) A scalar scipp variable (a value
+    with a unit) is interpreted as a target bin with, and an 'arange' covering the
+    min and max of the corresponding coordinate is created. (3) A custom coordinate,
+    given as a scipp variable with compatible unit. Typically this should have a single
+    dimension matching the target dimension.
+
+    When histogramming a dimension with an existing dimension-coord, the binning for
+    the dimension is modified, i.e., the input and the output will have the same
+    dimension labels.
+
+    When histogramming by non-dimension-coords, the output will have new dimensions
+    given by the names of these coordinates. These new dimensions replace the
+    dimensions the input coordinates depend on.
+
+    Parameters
+    ----------
+    x:
+        Input data.
+    arg_dict:
+        Dictionary mapping dimension labels to binning parameters.
+    **kwargs:
+        Mapping of dimension label to corresponding binning parameters.
+
+    Returns
+    -------
+    :
+        Histogrammed data.
+
+    See Also
+    --------
+    scipp.bin:
+        Creating binned data by binning instead of summing all contributions.
+    scipp.binning.make_histogrammed:
+        Lower level function for histogramming that does not automatically
+        replace/erase dimensions.
+
+    Examples
+    --------
+
+    Histogram a table by one of its coord columns, specifying (1) number of bins, (2)
+    bin width, or (3) actual binning:
+
+      >>> table = sc.data.table_xyz(100)
```

Comment:
It is unclear to most readers what exactly is in `table`. I think it would be better to define it here explicitly (doesn't even need `z`).